### PR TITLE
BrowserDebugProxy: unify debug metadata reading for PE and Webcil

### PR DESCRIPTION
**Pull request by lambdageek**
Factor common code out into a `MetadataDebugSummary` using an adapter to read from either `PEReader` or `WebcilReader`.

- [x] Implement `WebcilReader.ReadPdbChecksumDebugDirectoryData`

Contributes to #80807

### CODE REVIEW

**Review by thaystg — APPROVED** [review]

File: src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.cs [hunk 1]
```diff
@@ -297,6 +308,46 @@ private static MetadataReaderProvider DecodeEmbeddedPortablePdbDirectoryData(Blo
...
+    {
+        var algorithmName = ReadUtf8NullTerminated(reader);
+        byte[]? checksum = reader.ReadBytes(reader.RemainingBytes);
+        if (string.IsNullOrEmpty(algorithmName) || checksum == null || checksum.Length == 0)
+        {
+            throw new BadImageFormatException("Invalid PdbChecksum data format");
+        }
+
+        return MakePdbChecksumDebugDirectoryData(algorithmName, ImmutableArray.Create(checksum));
+
```
Comment (radical) [inline src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.cs:348]:
```suggestion
```

File: src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.cs [hunk 2]
```diff
@@ -240,6 +240,17 @@ private static CodeViewDebugDirectoryData MakeCodeViewDebugDirectoryData(Guid gu
         return (CodeViewDebugDirectoryData)mi.Invoke(new object[] { guid, age, path });
     }
 
+    private static PdbChecksumDebugDirectoryData MakePdbChecksumDebugDirectoryData(string algorithmName, ImmutableArray<byte> checksum)
+    {
+        var types = new Type[] { typeof(string), typeof(ImmutableArray<byte>) };
+        var mi = typeof(PdbChecksumDebugDirectoryData).GetConstructor(BindingFlags.NonPublic | BindingFlags.Instance, null, types, null);
```
Comment (radical) [inline src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.cs:246]: Would it make sense to cache this, if this is getting called per assembly?

File: src/mono/wasm/debugger/BrowserDebugProxy/IDebugMetadataProvider.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/IDebugMetadataProvider.cs:4]: 👍

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 4]
```diff
@@ -837,6 +837,73 @@ void AppendToBrowsable(Dictionary<string, DebuggerBrowsableState?> dict, CustomA
...
+                }
+                catch (BadImageFormatException)
+                {
+                    monoProxy.SendLog(sessionId, $"Warning: Unable to read debug information of: {name} (use DebugType=Portable/Embedded)", token);
+                }
+            }
+            else
+            {
+                var embeddedPdbEntry = entries.FirstOrDefault(e => e.Type == DebugDirectoryEntryType.EmbeddedPortablePdb);
+                if (embeddedPdbEntry.DataSize != 0)
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:897]: `embeddedPdbEntry` can be null here.

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 5]
```diff
@@ -837,6 +837,73 @@ void AppendToBrowsable(Dictionary<string, DebuggerBrowsableState?> dict, CustomA
         public override string ToString() => "TypeInfo('" + FullName + "')";
     }
 
+    /// <summary>
+    ///   Information we can extract directly from the assembly image using metadata readers
+    /// </summary>
+    internal sealed class MetadataDebugSummary
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:843]: nit: separate file. we get to use `#nullable enable` too then ;)

File: src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.cs [hunk 6]
```diff
@@ -297,6 +308,46 @@ private static MetadataReaderProvider DecodeEmbeddedPortablePdbDirectoryData(Blo
...
+            throw new BadImageFormatException("Could not read CodeView debug directory data", nameof(_stream));
+        }
+        unsafe
+        {
+            fixed (byte* p = buffer)
+            {
+                return DecodePdbChecksumDebugDirectoryData(new BlobReader(p, buffer.Length));
+            }
+        }
+
```
Comment (radical) [inline src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.cs:335]:
```suggestion
```

**Review by radical — APPROVED** [review]

File: src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.cs (hunk 2 above)
Comment (lambdageek) [reply src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.cs:246]: yea probably :D

Comment (lambdageek) [reply src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.cs]: Added some lazy helpers

File: src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.Reflection.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+namespace Microsoft.NET.WebAssembly.Webcil;
+
+
+public sealed partial class WebcilReader
+{
+
+    // Helpers to call into System.Reflection.Metadata internals
+    internal static class Reflection
+    {
+        private static Lazy<MethodInfo> _readUtf8NullTerminated = new Lazy<MethodInfo>(() =>
```
Comment (radical) [inline src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.Reflection.cs:22]: `runtime` coding style - static fields should have a `s_` prefix.
```suggestion
        private static readonly Lazy<MethodInfo> s_readUtf8NullTerminated = new Lazy<MethodInfo>(() =>
```

File: src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.Reflection.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            if (mi == null)
+            {
+                throw new InvalidOperationException("Could not find BlobReader.ReadUtf8NullTerminated");
+            }
+            return mi;
+        });
+
+        internal static string? ReadUtf8NullTerminated(BlobReader reader) => (string?)_readUtf8NullTerminated.Value.Invoke(reader, null);
+
+        private static Lazy<ConstructorInfo> _codeViewDebugDirectoryDataCtor = new Lazy<ConstructorInfo>(() =>
```
Comment (radical) [inline src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.Reflection.cs:34]:
```suggestion
        private static readonly Lazy<ConstructorInfo> _codeViewDebugDirectoryDataCtor = new Lazy<ConstructorInfo>(() =>
```

File: src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.Reflection.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            if (mi == null)
+            {
+                throw new InvalidOperationException("Could not find CodeViewDebugDirectoryData constructor");
+            }
+            return mi;
+        });
+
+        internal static CodeViewDebugDirectoryData MakeCodeViewDebugDirectoryData(Guid guid, int age, string path) => (CodeViewDebugDirectoryData)_codeViewDebugDirectoryDataCtor.Value.Invoke(new object[] { guid, age, path });
+
+        private static Lazy<ConstructorInfo> _pdbChecksumDebugDirectoryDataCtor = new Lazy<ConstructorInfo>(() =>
```
Comment (radical) [inline src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.Reflection.cs:47]:
```suggestion
        private static readonly Lazy<ConstructorInfo> _pdbChecksumDebugDirectoryDataCtor = new Lazy<ConstructorInfo>(() =>
```

**Review by radical — APPROVED** [review]

**Pull request merged**